website/blog/2023-06-12-multiple-same-components.md

```diff
@@ -0,0 +1,302 @@
+This method does all the heavy lifting for us. It replaces the default attribute names with the ones we provide. There is both the PHP and JS version of it.
+
+```php
+Components::render('heading', Components::props('intro', $attributes, [
+	'selectorClass' => 'intro',
+	'blockClass' => $componentClass
+]))
+```
+
+```js
+
+<HeadingEditor
+	{...props('intro', attributes, {
+		selectorClass: 'intro',
+		blockClass: componentClass,
+	})}
+/>
+
+```
+
+If you `var_dump` the props helper, you’ll notice the `prefix` key, which is built from the names of the blocks and components used hierarchically. For example, when looking at a regular Heading component in the Card block, the prefix will be:
+
+`cardCardHeading` - first is the Card block name, then the Card component name and finally Heading component name.
+
+The Heading component that is called Intro then has the prefix `cardCardIntro`. This prefix is then added to the attribute name, which then finally results in `cardCardHeadingSize` and `cardCardIntroSize` attributes, for example.
+
+While at first glance it seems strange to have this naming scheme, it actually allows us to figure out the hierarchy just by looking at the attribute name.
+
+## A step-by-step example
+
+The block we'll be creating as an example will be a block which we can use for some sort of comparison or listing pros and cons. To create it, we will need the following:
+- two `Heading` components
+- two `List` components
+
+The easiest way to create a new block is by using the boilerplate command
+
+`wp boilerplate blocks use-block --name=example`
+
+Once the new block is added to your project, rename it. Also, don’t forget to update all file names and imports in JS.
+
+We should start with `manifest.json`, where we define the components we'll use and set the default attributes. For now, just define the components and their keys:
+
+```json
+"components": {
+	"heading": "heading",
+	"lists": "lists",
+	"secondaryHeading": "heading",
+	"secondaryLists": "lists"
+}
+```
+
+The first section has the default key names, while the "duplicates" have different key names.
+
+After manifest, we can move to the JS part. As this is a fairly simple block without any advanced options or layouts, we need to add two `HeadingEditor` components and two `ListsEditor` components. To make styling easier, we can separate them in two `div` elements. When you’re finished, your code should look like this:
+
+```jsx
+import React from 'react';
+import { __ } from '@wordpress/i18n';
+import { HeadingEditor } from '../../../components/heading/components/heading-editor';
+import { ListsEditor } from '../../../components/lists/components/lists-editor';
+
+import { props, classnames, selector } from '@eightshift/frontend-libs/scripts';
+
+export const ComparisonEditor = ({ attributes, setAttributes }) => {
+	const {
+		blockClass,
+	} = attributes;
+
+	const comparisonPrimaryClass = classnames(
+		selector(blockClass, blockClass, 'primary'),
+	);
```

Comment:
```suggestion
	const comparisonPrimaryClass = selector(blockClass, blockClass, 'primary');
```
Don't need `classnames` if using only one selector 😄

Reply:
> Couple more 😄
> 
> Regarding the comment on (not) using the `selector` helper, honestly cannot remember the discussion, but it would feel a bit weird to have a mix of using and not using it. Of course, if others feel like no `selector` is the way to, I'm totally fine with that option.

I feel that the verbose method is better just for sake of a cleaner git history in case something gets added later on, just sayinn 😄